Used transaction in onFormSubmit.

jempajarillo opened the pull request:
Used Firebase Firestore’s runTransaction for reading data, ensuring that the latest state is always fetched.

1. Replaced direct document reads inside onFormSubmit with runTransaction to prevent stale data issues.
2. Passed the transaction object to business logic functions
3. Updated tests to reflect the changes—all tests have been modified accordingly and successfully passed. ✅

aris-i commented on src/types.ts:
```diff
@@ -26,6 +29,10 @@ export interface Action{
     message?: string
 }
 
+type txn1 = <T>(query: Query<T>)=> Promise<QuerySnapshot<T>>;
```
txnGet1

aris-i commented on src/types.ts:
```diff
@@ -26,6 +29,10 @@ export interface Action{
 }
 
+type txn1 = <T>(query: Query<T>)=> Promise<QuerySnapshot<T>>;
+type txn2 = <T>(documentRef: DocumentReference<T>)=> Promise<DocumentSnapshot<T>>;
```
txnGet2

jempajarillo commented:
@aris-i Done updating txn1 and txn2 type naming

jempajarillo commented:
@aris-i Done updating, please review again

jempajarillo commented:
@aris-i  Please check again 

1. Removed pagination
2. Removed unnecessary values for RunBusinessLogicStatus
3. Initialized runStatus value
4. Include all write inside the runTransaction of onFormSubmit
5. Put updateErrorMessage before calling distributeFn
6. Fixed tests to accommodate changes
7. Removed DistributeFn interface

aris-i commented on src/index.ts:
```diff
@@ -315,428 +317,457 @@ export async function onFormSubmit(
+      updateErrorMessage();
+
+      await distributeFnTransactional(txn, runStatus.logicResults);
+      await formRef.update({"@status": "finished"});
```
This update in status should happen after non-transactional logic results are distributed.

aris-i replied:
From this line to end of this transactional block should happen after the distribution of non transactional logic result

aris-i commented on src/index-utils.ts:
```diff
@@ -338,33 +242,24 @@ function getMatchingLogics(actionType: ActionType, modifiedFields: DocumentData,
-async function runLogic(action: Action, matchingLogics: LogicConfig[], nextPageMarkers: (object | undefined)[],
-  sharedMap: Map<string, any>, logicResults: LogicResult[]) {
-  for (let i = matchingLogics.length - 1; i >= 0; i--) {
+async function runLogic(txnGet: TxnGet, action: Action, matchingLogics: LogicConfig[], sharedMap: Map<string, any>,
```
I think this function is redundant with runBusinessLogics.  Let's transfer this code block to inside runBusinessLogics or put the for loop inside runBusinessLogics and just pass one logic here so that runLogic name will make more sense

jempajarillo commented:
@aris-i Please check again 

1. Moved formRefUpdate to finished after running non transactional
2. Removed runLogic and just copied whole code to runBusinesslogic

aris-i commented on src/index-utils.ts:
```diff
@@ -338,33 +242,37 @@ function getMatchingLogics(actionType: ActionType, modifiedFields: DocumentData,
       logicResults.push({...result, execTime, timeFinished: admin.firestore.Timestamp.now()});
-      if (status === "cancel-then-retry") {
-        return "cancel-then-retry";
+      if (result.status === "cancel-then-retry") {
```
Let's remove cancel-then-retry since this will be solved by transactional

aris-i commented on src/index.ts:
```diff
@@ -315,413 +317,183 @@ export async function onFormSubmit(
-          if (page === 0) {
-            await formRef.update({"@status": "finished"});
-          }
+      if (runStatus.status === "cancel-then-retry") {
```
Let's remove cancel-then-retry

aris-i commented on src/index.ts:
```diff
@@ -315,413 +317,183 @@ export async function onFormSubmit(
+      await distributeFnTransactional(txn, runStatus.logicResults);
+      const distributeLogicResultsEnd = performance.now();
+      const distributeLogicResultPerf: LogicResult = {
+        name: "distributeLogicResults",
```
let's rename this to distributeTransactionalLogicResults.  Let's also measure performrnace of distributeNonTransactionalLogicResults

aris-i commented on src/index.ts:
```diff
@@ -315,413 +317,183 @@ export async function onFormSubmit(
+      logicResults.push(distributeLogicResultPerf);
+    });
+
+    if (runStatus.status === "cancel-then-retry") {
```
Let's remove cancel-then-retry

aris-i commented on src/index.ts:
```diff
@@ -315,413 +317,183 @@ export async function onFormSubmit(
       return;
     }
 
+    await distributeNonTransactionalLogicResults(runStatus.logicResults, docPath);
```
let's measure performance of distributeNonTransactionalLogicResult

aris-i reviewed (changes requested):
Please see comments
